<a href="https://colab.research.google.com/github/linneverh/MachineLearning/blob/Assignment/assignments/Part%204/Assignment_4_2024.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### DSML investigation

You are part of the Suisse Impossible Mission Force, or SIMF for short. You need to uncover a rogue agent that is trying to steal sensitive information.

Your mission, should you choose to accept it, is to find that agent before stealing any classified information. Good luck!

# Assignement part four

#### Identifying the suspects' credit score
We received informations that the rogue agent has a *good* credit score.

Our spies at SIMF have managed to collect financial information relating to our suspects as well as a training dataset.

Create a Neural Network over the training dataset `df` to identify which of the suspects have a *"Good"* `Credit_Mix`.


## Getting to know our data

* `Age`: a user's age

* `Occupation`: a user's employment field

* `Annual_Income`: a user's annual income

* `Monthly_Inh_Salary`: the calculated salary received by a given user on a monthly basis

* `Num_Bank_Accounts`: the number of bank accounts possessed by a given user

* `Num_Credit_Cards`: the number of credit cards a given user possesses

* `Interest_Rate`: The interest rate on those cards (if multiple then it's the average)

* `Num_of_Loans`: The number of loans of each user

* `Delay_from_due_date`: payment tardiness of user

* `Num_of_Delayed_Payment`: the count of delayed payments

* `Changed_Credit_Limit`: NaN

* `Num_Credit_Inquiries`: NaN

* `Credit_Mix`: The user's credit score

* `Outsting_Debt`: Outstanding debt

* `Credit_Utilization_Ratio`: the percentage of borrowed money over borrowing allowance

* `Payment_of_Min_Amount`: does the user usually pay the minimal amount (categorical)

* `Total_EMI_per_month`: Monthly repayments to be made

* `Amount_invested_monthly`: The amount put in an investment fund by the user on a monthly basis

* `Payment_Behaviour`: the user's payment behavior (categorical)

* `Monthly_Balance`: The user's end of the month balance

* `AutoLoan`: If the user has an active loan for their vehicle

* `Credit-BuilderLoan`: If the user has a loan to increase their credit score

* `DebtConsolidationLoan`, `HomeEquityLoan`, `MortgageLoan`, `NotSpecified`, `PaydayLoan`, `PersonalLoan`, `StudentLoan`: different types of loans (categorical features)


In [20]:
# Import required packages

from sklearn.preprocessing import LabelEncoder, OneHotEncoder
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.preprocessing import MinMaxScaler

%matplotlib inline

In [53]:
df = pd.read_csv("https://raw.githubusercontent.com/michalis0/MGT-502-Data-Science-and-Machine-Learning/refs/heads/main/assignments/Part%204/data/train_classification.csv", index_col='Unnamed: 0').dropna()
suspects = pd.read_csv("https://raw.githubusercontent.com/michalis0/MGT-502-Data-Science-and-Machine-Learning/refs/heads/main/assignments/Part%204/data/suspects.csv", index_col='Unnamed: 0').dropna()
suspects.rename(columns={"Payment_Behaviour": "le_Payment_Behaviour", "Payment_of_Min_Amount": "le_Payment_of_Min_Amount"}, inplace=True)

In [50]:
display(df.head())
print(df.shape)
display(suspects.head())
print(suspects.shape)

Age Occupation  Annual_Income  Monthly_Inh_Salary  Num_Bank_Accounts  \
0   23  Scientist       19114.12         1824.843333                  3   
1   24  Scientist       19114.12         1824.843333                  3   
3   24  Scientist       19114.12         4182.004291                  3   
5   28    Teacher       34847.84         3037.986667                  2   
8   35   Engineer      143162.64         4182.004291                  1   

   Num_Credit_Card  Interest_Rate  Num_of_Loan  Delay_from_due_date  \
0                4              3            4                    3   
1                4              3            4                    3   
3                4              3            4                    4   
5                4              6            1                    3   
8                5              8            3                    8   

   Num_of_Delayed_Payment  ...  Monthly_Balance  AutoLoan Credit-BuilderLoan  \
0                       7  ...       186.266702         1                  1   
1                       9  ...       361.444004         1                  1   
3                       5  ...       343.826873         1                  1   
5                       3  ...       303.355083         0                  1   
8                    1942  ...       854.226027         2                  0   

   DebtConsolidationLoan  HomeEquityLoan MortgageLoan  NotSpecified  \
0                      0               1            0             0   
1                      0               1            0             0   
3                      0               1            0             0   
5                      0               0            0             0   
8                      0               0            0             1   

   PaydayLoan PersonalLoan  StudentLoan  
0           0            1            0  
1           0            1            0  
3           0            1            0  
5           0            0            0  
8           0            0            0  

[5 rows x 29 columns]

(29223, 29)


Age  Annual_Income  Monthly_Inh_Salary  Num_Bank_Accounts  \
0  23.0       19114.12         1824.843333                3.0   
1  24.0       19114.12         1824.843333                3.0   
3  24.0       19114.12         4182.004291                3.0   
5  28.0       34847.84         3037.986667                2.0   
8  35.0      143162.64         4182.004291                1.0   

   Num_Credit_Card  Interest_Rate  Num_of_Loan  Delay_from_due_date  \
0              4.0            3.0          4.0                  3.0   
1              4.0            3.0          4.0                  3.0   
3              4.0            3.0          4.0                  4.0   
5              4.0            6.0          1.0                  3.0   
8              5.0            8.0          3.0                  8.0   

   Num_of_Delayed_Payment  Changed_Credit_Limit  ...  Occupation_Journalist  \
0                     7.0                 11.27  ...                    0.0   
1                     9.0                 13.27  ...                    0.0   
3                     5.0                 11.27  ...                    0.0   
5                     3.0                  5.42  ...                    0.0   
8                  1942.0                  7.10  ...                    0.0   

   Occupation_Lawyer  Occupation_Manager  Occupation_Mechanic  \
0                0.0                 0.0                  0.0   
1                0.0                 0.0                  0.0   
3                0.0                 0.0                  0.0   
5                0.0                 0.0                  0.0   
8                0.0                 0.0                  0.0   

   Occupation_MediaManager  Occupation_Musician  Occupation_Scientist  \
0                      0.0                  0.0                   1.0   
1                      0.0                  0.0                   1.0   
3                      0.0                  0.0                   0.0   
5                      0.0                  0.0                   0.0   
8                      0.0                  0.0                   0.0   

   Occupation_Teacher  Occupation_Writer  userID  
0                 0.0                0.0  317991  
1                 0.0                0.0  241892  
3                 1.0                0.0  303376  
5                 0.0                0.0  761992  
8                 0.0                0.0  373318  

[5 rows x 43 columns]

(714, 43)


In [23]:
df["Credit_Mix"].unique()

array(['Good', 'Standard', 'Bad'], dtype=object)

# 1. Preparing the data
## 1.1 Data cleaning
Consider the dataset loaded into the DataFrame `df` which is *train_classification.csv*. We aim to preprocess this data for model training.

Begin by encoding the categorical variables:
- Apply One-Hot Encoding to `Occupation`
- Apply Label Encoding to `Payment_of_Min_Amount` and `Payment_Behaviour`

*Note: To clearly distinguish your newly encoded columns, especially for label encoding, consider renaming them with a prefix. Please, use `le_Payment_of_Min_Amount` and `le_Payment_Behaviour` for the label-encoded new columns.*

In [24]:
# Your code here
#label encode the categorical variables
le = LabelEncoder()

df['le_Payment_of_Min_Amount'] = le.fit_transform(df['Payment_of_Min_Amount'])
df['le_Payment_Behaviour'] = le.fit_transform(df['Payment_Behaviour'])

# One-hot encode Occupation
ohe = OneHotEncoder(sparse_output=False)

Occupation_encoded = ohe.fit_transform(df[['Occupation']])
df[ohe.get_feature_names_out(['Occupation'])] = Occupation_encoded


#restitute & remove original to not have redundancy
df_encoded = df
df_encoded.drop(columns=['Payment_of_Min_Amount', 'Payment_Behaviour', 'Occupation'], inplace=True)
df_encoded.head()

Age  Annual_Income  Monthly_Inh_Salary  Num_Bank_Accounts  Num_Credit_Card  \
0   23       19114.12         1824.843333                  3                4   
1   24       19114.12         1824.843333                  3                4   
3   24       19114.12         4182.004291                  3                4   
5   28       34847.84         3037.986667                  2                4   
8   35      143162.64         4182.004291                  1                5   

   Interest_Rate  Num_of_Loan  Delay_from_due_date  Num_of_Delayed_Payment  \
0              3            4                    3                       7   
1              3            4                    3                       9   
3              3            4                    4                       5   
5              6            1                    3                       3   
8              8            3                    8                    1942   

   Changed_Credit_Limit  ...  Occupation_Entrepreneur Occupation_Journalist  \
0                 11.27  ...                      0.0                   0.0   
1                 13.27  ...                      0.0                   0.0   
3                 11.27  ...                      0.0                   0.0   
5                  5.42  ...                      0.0                   0.0   
8                  7.10  ...                      0.0                   0.0   

   Occupation_Lawyer  Occupation_Manager  Occupation_Mechanic  \
0                0.0                 0.0                  0.0   
1                0.0                 0.0                  0.0   
3                0.0                 0.0                  0.0   
5                0.0                 0.0                  0.0   
8                0.0                 0.0                  0.0   

   Occupation_MediaManager  Occupation_Musician  Occupation_Scientist  \
0                      0.0                  0.0                   1.0   
1                      0.0                  0.0                   1.0   
3                      0.0                  0.0                   1.0   
5                      0.0                  0.0                   0.0   
8                      0.0                  0.0                   0.0   

   Occupation_Teacher  Occupation_Writer  
0                 0.0                0.0  
1                 0.0                0.0  
3                 0.0                0.0  
5                 1.0                0.0  
8                 0.0                0.0  

[5 rows x 43 columns]

 After encoding, integrate the newly encoded columns back into a new DataFrame named `df_encoded`, and remove the original columns `Occupation`, `Payment_of_Min_Amount`, and `Payment_Behaviour` to avoid redundancy.

Finally, display the first few rows of `df_encoded` to verify that the encodings are correctly implemented. This prepared DataFrame will be used for subsequent model training.

## 1.2 Dataset splitting and rescaling

To effectively train and validate our model, it is crucial to properly prepare and partition the data. Follow these steps to preprocess and split the dataframe df_encoded into training and test subsets, ensuring that our model can generalize well to new data:

- Set `X` as all columns except `Credit_Mix` and `y` as the dependent feature `Credit_Mix`.
- Apply manually Label Encoding to `y` using the function `.map`and encoding with the following : Good=2, Standard=1, Bad=0.
- Use a `random_state` of 42 to split `X` and the encoded `y` into training (80%) and test sets (20%).
- Normalize `X` using `MinMaxScaler()`.

In [26]:
# Define features and target variable
X = df_encoded.drop(columns='Credit_Mix')
y = df_encoded['Credit_Mix']

#manually label encoding
mapping = {'Good': 2, 'Standard': 1, 'Bad': 0}
y= y.map(mapping)

#train vs test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, shuffle=True)

#scaler
scaler = MinMaxScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

### 1.2.2 Final touches
Convert the features to torch tensors of type `torch.float` and the labels (dependent variables) to torch tensors of type `torch.long`.

In [33]:
# Convert features and labels to PyTorch tensors
X_train = torch.tensor(X_train, dtype=torch.float)
X_test = torch.tensor(X_test, dtype=torch.float)
y_train = torch.tensor(y_train, dtype=torch.long)
y_test = torch.tensor(y_test, dtype=torch.long)

<ipython-input-33-6cf110d1264a>:4: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  X_train = torch.tensor(X_train, dtype=torch.float)
<ipython-input-33-6cf110d1264a>:5: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  X_test = torch.tensor(X_test, dtype=torch.float)
<ipython-input-33-6cf110d1264a>:6: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  y_train = torch.tensor(y_train, dtype=torch.long)
<ipython-input-33-6cf110d1264a>:7: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor

In [34]:
print(X_train.size(), y_train.size())

torch.Size([23378, 42]) torch.Size([23378])


# 2 Model preparation:

## 2.1 Define a Neural network model and instantiate it.
Define your neural network model as a class in PyTorch, extending from nn.Module. In the `__init__` method, initialize a linear layer using `nn.Linear()` with specified input and output sizes, and set up `nn.ReLU()` for activation. Implement the forward method to describe how data passes through this layer during the network's forward computation.

Set the following parameters:
* `hidden layer` : 1
* `activation function` : ReLU

In [35]:
# Your code here
class Net(nn.Module):
    def __init__(self, D_in, H1, D_out):
        super(Net, self).__init__()

        self.linear1 = nn.Linear(D_in, H1)        # Linear transformation for hidden layer
        self.linear2 = nn.Linear(H1, D_out)       # Linear transformation for output layer
        self.activation = nn.ReLU()               # Activation function for hidden layer

    def forward(self, x):
        y_pred = self.activation(self.linear1(x))   # Hidden layer: linear transformation + ReLU
        y_pred = self.linear2(y_pred)               # Output layer: linear transformation
        return y_pred

Set `D_in` to the number of features in `X_train` and `D_out` to the number of target variables in `y_train`, then print these dimensions to verify their values.

In [41]:
D_in, D_out = X_train.shape[1], 3


Initialize the `Net` model with the specified input size `D_in`, 150 hidden units, and output size `D_out`.

In [43]:
# Model with 150 neurons
model1 = Net(D_in, 150, D_out)

Let's calculate now how many parameters we have in the model.

In [44]:
# Calculate the number of parameters in the model
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)

print(f"Number of parameters in model1: {count_parameters(model1)}")


Number of parameters in model1: 6903


**Q1. How many parameters does your model have ?**

*Note: Enter an integer (e.g. 355)* - 6903

## 2.2 Finding the best model

Determine the optimal hyper-parameters for your model from the options listed below:

* `criterion` : [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)
* `optimizer` : [Stochastic Gradient Descent (SGD)](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)
* `Epochs`: Test with **150**, **250**, **500**, and **1000** epochs.
* `Learning Rate`: Experiment with learning rates of **0.00005**, **0.001**, **1**, and **10**.

**Evaluation**: Assess your model's performance by measuring its accuracy on the test set.

*Note: Run the code `torch.manual_seed(42)` to ensure consistency across all experiments.*

In [45]:
torch.manual_seed(42)   # Set the seed for reproducibility

### 2.2.1 Automatically Tuning Hyperparameters

In this section, you will automate the process of testing different hyperparameter combinations using a loop, as demonstrated in the lab.

Begin by defining the ranges for the hyperparameters you want to explore:
- **Epochs**: Test with **150**, **250**, **500**, and **1000** epochs.
- **Learning Rate**: Experiment with learning rates of **0.00005**, **0.001**, **1**, and **10**.


In [46]:
# Begin by defining the ranges for the hyperparameters you want to explore:
epochs_list = [150, 250, 500, 1000]
learning_rates = [0.00005, 0.001, 1, 10]


Next, create a loop that iterates over the list of learning rates, with an inner loop iterating over the list of number of epochs. Inside the inner loop, initialize (define) the model, and also define the optimizer and the loss function (criterion). Then train the model as demonstrated in the lab, and after the training evaluate its performance to obtain the test accuracy for each model. Ensure that you also display the test loss, as you will need it to answer the upcoming questions.

Here’s a reminder of the criterion and optimizer you should use:
- **Criterion**: [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)
- **Optimizer**: [Stochastic Gradient Descent (SGD)](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)

**Hint**: Make sure to define the criterion and optimizer at every iteration (inside the inner loop), otherwise the model might retain previous training states and produce biased results.

(Optional) You can enhance the code by storing the best model along with its best_accuracy and best_params. This way, you'll have an automatic evaluation at the end to identify the best-performing model.



In [47]:
best_accuracy = 0
best_params = {}
best_model = None

for lr in learning_rates:
  for epochs in epochs_list:
    # Initialize the model, criterion, and optimizer
    model = Net(D_in, 150, D_out)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(model.parameters(), lr=lr)

    # Training loop
    for epoch in range(epochs):
      optimizer.zero_grad()
      outputs = model(X_train)
      loss = criterion(outputs, y_train)
      loss.backward()
      optimizer.step()

    # Evaluation
    with torch.no_grad():
      outputs_test = model(X_test)
      _, predicted = torch.max(outputs_test, 1)
      correct = (predicted == y_test).sum().item()
      accuracy = correct / len(y_test)
      loss_test = criterion(outputs_test, y_test).item()

    print(f"Learning Rate: {lr}, Epochs: {epochs}, Test Accuracy: {accuracy:.4f}, Test Loss: {loss_test:.4f}")

    # Store the best model
    if accuracy > best_accuracy:
      best_accuracy = accuracy
      best_params = {'learning_rate': lr, 'epochs': epochs}
      best_model = model

print(f"Best Accuracy achieved: {best_accuracy:.4f} with parameters: {best_params}")


Learning Rate: 5e-05, Epochs: 150, Test Accuracy: 0.3092, Test Loss: 1.1037
Learning Rate: 5e-05, Epochs: 250, Test Accuracy: 0.2821, Test Loss: 1.1161
Learning Rate: 5e-05, Epochs: 500, Test Accuracy: 0.3904, Test Loss: 1.0894
Learning Rate: 5e-05, Epochs: 1000, Test Accuracy: 0.4236, Test Loss: 1.0915
Learning Rate: 0.001, Epochs: 150, Test Accuracy: 0.4346, Test Loss: 1.0847
Learning Rate: 0.001, Epochs: 250, Test Accuracy: 0.4559, Test Loss: 1.0765
Learning Rate: 0.001, Epochs: 500, Test Accuracy: 0.4546, Test Loss: 1.0689
Learning Rate: 0.001, Epochs: 1000, Test Accuracy: 0.4546, Test Loss: 1.0522
Learning Rate: 1, Epochs: 150, Test Accuracy: 0.8019, Test Loss: 0.4318
Learning Rate: 1, Epochs: 250, Test Accuracy: 0.8308, Test Loss: 0.3732
Learning Rate: 1, Epochs: 500, Test Accuracy: 0.8378, Test Loss: 0.3542
Learning Rate: 1, Epochs: 1000, Test Accuracy: 0.8429, Test Loss: 0.3465
Learning Rate: 10, Epochs: 150, Test Accuracy: 0.4546, Test Loss: 2.6760
Learning Rate: 10, Epochs: 2

### 2.2.3 Questions

**Q2. What is the test accuracy when we train the model with a learning rate of 0.001 and for 150 epochs? Round your answer to 2 decimal points, e.g., 0.25.** 0.43


**Q3. When using 1000 epochs, which learning rate results in the highest test accuracy?**

*Note: Select among the following answers*


**Q4. Is BCELoss a suitable alternative to CrossEntropyLoss for our dataset?**

*Hint: Consider the unique values in the Credit_Mix output variable when answering.*

### 3. Predict on the Suspects Dataset

Now it's time to use your trained model to make predictions on the suspects dataset!

Please retrain on the full dataset.

Use the following parameters for the model:
- **Hidden layer**: 1 hidden layer with 150 neurons
- **Output layer**: 3 neurons for classification
- **Optimizer**: [Stochastic Gradient Descent (SGD)](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)
- **Criterion**: [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)
- **Iterations**: 1000
- **Learning rate**: 1.0

Ensure consistency by setting the manual seed with `torch.manual_seed(42)` before training.

In [69]:
# prompt: check which columns not in df_encoded and in suspects

# Assuming df_encoded and suspects are defined as in the provided code.

cols_not_in_encoded = []
for col in suspects.columns:
    if col not in df_encoded.columns:
        cols_not_in_encoded.append(col)

print("Columns in 'suspects' but not in 'df_encoded':", cols_not_in_encoded)


Columns in 'suspects' but not in 'df_encoded': ['userID']


In [60]:
torch.manual_seed(42)   # Set the seed for reproducibility

In [71]:
# Define features and target variable
X = df_encoded.drop(columns='Credit_Mix')
y = df_encoded['Credit_Mix']

#manually label encoding
mapping = {'Good': 2, 'Standard': 1, 'Bad': 0}
y= y.map(mapping)


#scaler
scaler = MinMaxScaler()
scaler.fit(X)
X_train = scaler.transform(X)
X_test = suspects.drop(columns=['userID'])
X_test = scaler.transform(X_test)

ValueError: The feature names should match those that were passed during fit.
Feature names must be in the same order as they were in fit.


In [61]:
D_in, D_out = X.shape[1], 3

In [66]:
# Retrain the model on the full dataset with specified parameters
model_full = Net(D_in, 150, D_out)
criterion_full = nn.CrossEntropyLoss()
optimizer_full = optim.SGD(model_full.parameters(), lr=1.0)

# Prepare the full dataset
X_full = torch.tensor(scaler.fit_transform(X), dtype=torch.float)
y_full = torch.tensor(y.map(mapping), dtype=torch.long)

# Training loop for the full dataset
for epoch in range(1000):
  optimizer_full.zero_grad()
  outputs_full = model_full(X_full)
  loss_full = criterion_full(outputs_full, y_full)
  loss_full.backward()
  optimizer_full.step()

# Align the suspects DataFrame columns with those used during training
X_suspects = suspects.drop(columns=['userID'])  # Remove the 'userID' column
# Ensure the order of columns in X_suspects matches the order in X_full
X_suspects = X_suspects[X.columns] # This assumes that X is the original DataFrame used for training

X_suspects = torch.tensor(scaler.transform(X_suspects), dtype=torch.float)

# Make predictions on the suspects dataset
with torch.no_grad():
    outputs_suspects = model_full(X_suspects)
    _, predicted_suspects = torch.max(outputs_suspects, 1)

# Print or use the predictions
predicted_suspects

tensor([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 2, 2, 2, 2, 2, 1, 1, 1, 1, 2, 2, 2,
        2, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 2, 2,
        2, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 2, 2, 0, 1, 1, 1, 1,
        1, 2, 1, 1, 1, 0, 0, 0, 2, 2, 1, 1, 2, 2, 2, 2, 1, 1, 1, 0, 0, 0, 1, 1,
        1, 1, 1, 2, 2, 2, 2, 2, 0, 0, 0, 0, 0, 0, 0, 2, 2, 2, 0, 0, 1, 2, 1, 1,
        1, 0, 0, 0, 0, 0, 0, 2, 1, 1, 1, 2, 2, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0,
        1, 0, 0, 1, 1, 2, 2, 2, 2, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0,
        0, 0, 2, 2, 2, 2, 1, 2, 2, 0, 0, 0, 2, 2, 2, 0, 0, 0, 1, 1, 2, 2, 2, 2,
        2, 0, 0, 0, 2, 2, 2, 2, 1, 1, 1, 1, 0, 0, 0, 0, 1, 2, 2, 2, 2, 2, 2, 2,
        2, 2, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 2, 2, 2, 1, 1, 1, 0, 0, 2, 2, 1,
        1, 2, 1, 2, 2, 0, 0, 0, 0, 0, 0, 1, 2, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 1,
        1, 1, 1, 1, 1, 1, 2, 2, 2, 1, 1, 1, 0, 0, 0, 0, 2, 2, 2, 0, 0, 0, 1, 1,
        1, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1,


Now, train your model on the training dataset just as you did in section 2.2.1, but you don't have to loop over all the hyper parameters. Ensure that you use the correct number of epochs specified earlier.

In [ ]:
# Your code here

Before making predictions, confirm that the feature column names in the `suspects` dataset match those expected by the model, particularly ensuring the X features correspond accurately.

In [ ]:
# Your code here

Then scale your dataset and convert it into a torch tensor of dtype float, similar to the preprocessing done for the training set in section 1.2.

In [ ]:
#Your code here

Make predictions using the trained model and assign the predicted credit score to each user. Ensure to do the following encoding is used for the predicted categories:
* `0` corresponds to bad credit score,
* `1` corresponds to standard credit score,
* `2` corresponds to good credit score.

Use the predictions to add a new column `credit_score` in the `suspects` dataframe that maps the predicted numerical values to the respective credit score categories.

In [ ]:
# Your code here



As mentioned earlier, we believe the suspect had a "Good" credit score. Review the predictions made by the model, then display how many suspects were categorized under each credit score, and extract the `userID`s of those with a "Good" credit score.

In [ ]:
# Your code here

**Q5.Which of the following suspects have a "good" credit mix according to your model's predictions?**


## Your investigation is progressing effectively, and the list of suspects is narrowing down.

**Don't forget to answer the quiz and submit your code on Moodle before the end of the deadline.**